### ISSUE TITLE
Config example for OIDC server in a development env

### ISSUE THREAD

**Issue by lauracowen** [OpenLiberty/docs#676]
From Adam Yoho:

```
<server>
    <featureManager>
        <!-- The jdbc-4.0 feature is needed if using a database to store OAuth clients, as shown in this example. A database is the recommended storage type to use in production. -->
        <feature>jdbc-4.0</feature>
        <feature>openidConnectServer-1.0</feature>
    </featureManager>

    <openidConnectProvider
        id="OP"
        oauthProviderRef="OAuth"
        jwkEnabled="true" />

    <oauthProvider id="OAuth"  tokenFormat="mpjwt" >
        <!-- This indicates that the OAuth provider will use a database to store OAuth clients instead of manually configuring them in the server configuration. -->
        <databaseStore
            dataSourceRef="OAuthFvtDataSource"
            user="dbuser"
            password="dbpwd" />
    </oauthProvider>

    <!-- This example uses Derby as the configured database, but any database type that is supported by Liberty can be used. -->
    <dataSource
        id="OAuthFvtDataSource"
        jndiName="jdbc/OAuth2DB"
        jdbcDriverRef="DerbyEmbedded"
    >
        <properties.derby.embedded databaseName="memory:oAuthDB" createDatabase="create" />
    </dataSource>
    <jdbcDriver id="DerbyEmbedded">
        <library>
            <fileset dir="${server.config.dir}/derby" includes="derby.jar" />
        </library>
    </jdbcDriver>

    <oauth-roles>
        <authenticated>
            <special-subject type="ALL_AUTHENTICATED_USERS" />
        </authenticated>
    </oauth-roles>

    <!-- An OpenID Connect server needs a user registry to authenticate users. This example points to an external LDAP server that maintains the user registry. -->
    <ldapRegistry id="bluepages" realm="w3" host="bluepages.ibm.com" port="389" ignoreCase="true" baseDN="o=ibm.com" ldapType="IBM Tivoli Directory Server" >
        <idsFilters
            userFilter="(&amp;(emailAddress=%v)(objectclass=person))"
            groupFilter="(&amp;(cn=%v)(objectclass=groupOfUniqueNames))"
            userIdMap="*:emailAddress"
            groupIdMap="*:cn"
            groupMemberIdMap="groupOfUniqueNames:uniquemember" />
    </ldapRegistry>

    <!-- In order to communicate with OpenID Connect clients over TLS, the trust store used by the OpenID Connect server must contain the appropriate SSL certificates of the respective OpenID Connect clients. -->
    <sslDefault sslRef="DefaultSSLSettings" />
    <ssl
        id="DefaultSSLSettings"
        keyStoreRef="myKeyStore"
        trustStoreRef="myTrustStore"
        clientAuthenticationSupported="true" />
    <keyStore
        id="myKeyStore"
        password="Liberty"
        type="jks"
        location="${server.config.dir}/commonBasicKeyStore.jks" />
    <keyStore
        id="myTrustStore"
        password="Liberty"
        type="jks"
        location="${server.config.dir}/commonTrustStore.jks" />

    <httpEndpoint id="defaultHttpEndpoint" host="*" httpPort="29080" httpsPort="29443" />

</server>
```
This may be more full (at the very least, it shouldn't include the `server` elements) than we should show but add it and then get Alasdair to review the PR.

**Comment by Charlotte-Holt** [comment]
This should be a simple example for use in development (not for use in production).

**Comment by lauracowen** [comment]
Replacement for the example above based on the server.xml from Bruce...(lay out in sections as in the JDBC and social media login features; format the XML with consistent indenting).

config example for OIDC server (for development purposes):

- assumes using basic registry which is available for purposes of development/test (if using another method of authentication (e.g. LDAP), replace the basicRegistry section with <LDAP>[link to LDAP feature where there'll be example config] config.
- uses localStore to store client data and token status, which means all client data and token status is held in memory (only for test/development). If you use a database, replace the localStore section with <JDBC>[link to JDBC feature which has example config] database configuration; when using a database, clients are not specified in server.xml.Instead they are added to the database through the registration endpoint. If you reconfigure the server, localStore might be cleared.

```
    <keyStore id="defaultKeyStore" password="keyspass" />  

    <basicRegistry id="basic" realm="customRealm">
        <user
          name="demouser"
          password="demopassword" />
        <user
          name="demouser2"
          password="demopassword2" />
         <group name="users">
             <member name="demouser"/>
             <member name="demouser2" />
        </group>
    </basicRegistry>

    <openidConnectProvider id="OP" 
        oauthProviderRef="OAuth" 
        signatureAlgorithm="RS256" keyStoreRef="defaultKeyStore" 
        jwkEnabled="true">
    </openidConnectProvider>

    <oauthProvider id="OAuth" tokenFormat="mpjwt" >
        <localStore>
          <client displayname="RP" enabled="true"
                name="RP" secret="thesecret" 
                scope="openid profile email"
                preAuthorizedScope="openid profile email">      
                <redirect>https://localhost:19443/oidcclient/redirect/RP</redirect>
          </client>    
        </localStore>
    </oauthProvider>
 
    <oauth-roles>
        <authenticated>
            <special-subject type="ALL_AUTHENTICATED_USERS" />  
        </authenticated>   
    </oauth-roles>
```
---
Example of adding clients (when using localStore only) [find out what this means - when do you need to add clients?]:

```
    <oauth-roles>
        <authenticated>
            <special-subject type="ALL_AUTHENTICATED_USERS" />  
        </authenticated>     
        <clientManager>            
            <user name="testuser" />
            <group name="oidcadmin" />            
        </clientManager>  
    </oauth-roles>

---

example of using a database as the store:

```
    <oauthProvider id="OAuth" tokenFormat="mpjwt" >
        <localStore>
          <client displayname="RP" enabled="true"
                name="RP" secret="thesecret" 
                scope="openid profile email"
                preAuthorizedScope="openid profile email">      
                <redirect>https://localhost:19443/oidcclient/redirect/RP</redirect>
          </client>    
        </localStore>
   <databaseStore dataSourceRef="OIDC_DBDS4" />
    </oauthProvider>
```

and configure the datasource with authentication <link to JDBC feature for example>. To add clients when using a database, users in the clientManager role can add/modify clients by accessing the registration endpoint.

**Comment by lauracowen** [comment]
Moving this back to In Progress as I don't see any of these updates. I'm not sure why it's in SME Review if it's still being worked on. There's no problem shifting issues back and forth between columns if there are significant changes to make after an SME review.

**Comment by ManasiGandhi** [comment]
> 
> 
> Moving this back to In Progress as I don't see any of these updates. I'm not sure why it's in SME Review if it's still being worked on. There's no problem shifting issues back and forth between columns if there are significant changes to make after an SME review.

I was working on this feature with Adam Yoho (he put his suggestions in Slack https://ibm-cloud.slack.com/archives/DL551F6FL/p1581714784003000)and hence put it in SME review. Now will be working with the new example from Bruce.

**Comment by ManasiGandhi** [comment]
Requesting @Charlotte-Holt for a peer review of the built draft https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#openidConnectServer-1.0.html

**Comment by Charlotte-Holt** [comment]
Peer review feedback:
- "The OpenID Connect Provider(OP) feature can be configured to manage user authentication for OpenID Connect(OIDC) clients and microservices during development." - In this sentence, there should be spaces on both sides of the acronyms that are in parentheses (e.g., "OpenID Connect Provider (OP)").
- I'd suggest updating the second sentence to "The OP feature can work directly with a user registry, such as LDAP, or act as an identity broker and delegate to another security provider." to add commas to offset "such as LDAP". 
- "In production, the application uses a cloud-hosted SSO provider." - Could you be more specific in this sentence? I'm not sure what application you're referring to. 
- Add the OP Open Liberty feature and any other needed features to the server.xml file." - I'm not sure you need this sentence. The **Enabling this feature** section mentions this. 
- "Any user that is authorized to use OIDC must also be mapped to the authenticated OAuth-role." - I don't think that you need the hyphen unless you make it monospaced and are referring to the particular element, but I think the element is `oauth-roles`? Check on that one. 
- "The following example assumes that you are using a basic registry, which is available for development and test purposes." - add this comma before "which"
- Update "The example uses `localStore` to store client data and token status, which means all client data and token status is held in memory, for test/development purposes." to "The example uses `localStore` to store client data and token status, which means all client data and token status is held in memory for test and development purposes."
- I'm not sure what <keyStore id="defaultKeyStore" password="keyspass" /> is doing in this section exactly? Maybe supposed to be part of the example? 
- I think "datasource" should be "data source" in this context.
- Maybe look into updating "To add clients when using a database, users in the clientManager role can add or modify clients by accessing the registration endpoint." to "When you connect to a database, users in the clientManager role can add or modify clients by accessing the registration endpoint." - This makes more sense to me, but check it's accuracy.
- The second and third examples I think could use some more context rather than just the short lead-in (e.g., "Example of adding clients with `localStore`:". These lead-ins also need to be in complete sentences. Can you review with your SME to add a sentence or two about what the example is doing and when someone would use it? 
- I think there may be a typo in the third config snippet. It has "```" in the code block on the last line. 
- Last thing, I think the top part of the examples section is a bit choppy with how there are multiple lines with just one or two sentences. Can you try to make this part a little more cohesive and smooth?

**Comment by ayoho** [comment]
I'm a little confused about the config snippets at the bottom of the article. The large snippet seems to have a good lead-in and appropriate config. But the config snippet after the line that says "Example of adding clients with localStore:" is an `<oauth-roles>` snippet whose only difference from the config above it is that it has a `<clientManager>` stanza. So that config snippet is more for how to add users and groups to the `clientManager` role so that they can manage OAuth and OIDC clients rather than how to add clients with `localStore`.

The config snippet after the "Example of using a database as the store:" line shows an OAuth provider that uses `localStore` to store its clients, not a database.

Then the bottom of the article has "Configure the data source with authentication ." I assume that would be better suited to be paired with wherever we had an example of using a database as the client store. A sample config using a database with authentication is available in the description of this issue, if one is needed.

**Comment by lauracowen** [comment]
**Intro section** (at the top of the page)
- Change the opening introduction to the page to capture the bit about development vs production so:
  "OIDC is a protocol that helps developers authenticate users without the need to manage user credentials. You can configure Open Liberty to act as an OIDC provider. With this configuration, users need to authenticate once to access Open Liberty resources such as HTML, JavaServer Pages (JSP) files, and servlets. Users can also access resources in multiple Open Liberty servers that share Lightweight Third Party Authentication (LTPA) keys." becomes
  "OIDC is a protocol that helps developers authenticate users without the need to manage user credentials. You can configure Open Liberty to act as an OIDC provider. This is especially useful during development; when deployed into production in the cloud, apps typically use a cloud-hosted SSO provider instead. With this configuration, users need to authenticate once to access Open Liberty resources such as HTML, JavaServer Pages (JSP) files, and servlets. Users can also access resources in multiple Open Liberty servers that share Lightweight Third Party Authentication (LTPA) keys."
- Collapse the second and third sentences into a single paragraph - they don't need to be separate.

**Examples section**
- Don't abbreviate names of features. ie just call it the "OpenID Connect Provider feature" each time (not "OP feature").
- Though, actually, just remove the first and second paragraphs (from "The OpenID Connect Provider...." to "...the authenticated OAuth role." under the Examples heading. That info is now already covered at the top of the page.
- Insert a heading that captures the purpose of this first example config (see below).
- Then start the sub-section with "The following example shows how to create a basic configuration to run an OpenID Connect Provider. This is especially useful to use during development and testing prior to deploying to the cloud (where the application typically uses a cloud-based SSO provider instead). The example uses a simple basic registry, which is defined in the server configuration, to manage a single user:" and remove the paragraph "The following example assumes......with LDAP config.".
- Make "basic registry" into a link to the Application Security feature (it contains examples and explanations of basic registry configurations).
- Then have the first example here.
- Simplify the `basicRegistry` section of the example so that it only shows one user and no groups (as shown in the first example config in [https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#appSecurity-3.0.html](AppSecurity feature). We need this to be as minimal as possible; we don't need to show multiple users/groups here as that's not what we're demonstrating in this example.
- Then after the example, explain what the example shows: "The example configuration uses localStore to store client data and token status, which means all client data and token status is held in memory. This is fine for test and development purposes but if you were to reconfigure the server, the localStore might be cleared so it is not suitable to use for production purposes when a database should be used instead. (then explain what the settings are doing and where those values come from - otherwise it just looks like a bunch of values plucked from the ether.)"
- Then explain what the `openidConnectProvider` values show and where they come from.
- I don't think we need the third example (lead-in says "Example to add clients with localstore:") because that exact example is already part of the larger first example, which is described as using localstore.
- Can we also @ayoho just move the following clientManager lines into the "data source with authentication" example instead of it standing alone as it's not really an alternative configuration?
  ```
  <clientManager>
     <user name="testuser" />
     <group name="oidcadmin" />
  </clientManager>
  ```
  Then briefly explain what it does.
- So then we just have two examples but they each need their own sub-heading that captures what they're demonstrating. eg "Managing users with basic registry" and "Managing users with an authenticated database" maybe? @ayoho
- The indentations in the XML in the final example needs tidying up so that opening and closing tags match and everything's correctly nested. Also, the random `>` character should be on the previous line, not on a line on its own.

**Comment by ayoho** [comment]
- Moving the `<clientManager>` lines into the `<dataSource>` example is perfectly fine with me.
- Having just the two examples (local store and database) is also fine with me. I'll clarify, though, that the localStore/dataSource aren't for managing users - they're for OAuth clients to be used with the OpenID Connect Provider. So the first example would be "Managing OAuth clients with localStore" and the second would be "Managing OAuth clients with an authenticated database".

**Comment by ManasiGandhi** [comment]
@lauracowen I worked on your review.


    

- [x] Change the opening introduction to the page to capture the bit about development vs production so:

    "OIDC is a protocol that helps developers authenticate users without the need to manage user credentials. You can configure Open Liberty to act as an OIDC provider. With this configuration, users need to authenticate once to access Open Liberty resources such as HTML, JavaServer Pages (JSP) files, and servlets. Users can also access resources in multiple Open Liberty servers that share Lightweight Third Party Authentication (LTPA) keys." becomes
    "OIDC is a protocol that helps developers authenticate users without the need to manage user credentials. You can configure Open Liberty to act as an OIDC provider. This is especially useful during development; when deployed into production in the cloud, apps typically use a cloud-hosted SSO provider instead. With this configuration, users need to authenticate once to access Open Liberty resources such as HTML, JavaServer Pages (JSP) files, and servlets. Users can also access resources in multiple Open Liberty servers that share Lightweight Third Party Authentication (LTPA) keys."
    

- [x] Collapse the second and third sentences into a single paragraph - they don't need to be separate.

Examples section

    

- [x] Don't abbreviate names of features. ie just call it the "OpenID Connect Provider feature" each time (not "OP feature").

    

- [x] Though, actually, just remove the first and second paragraphs (from "The OpenID Connect Provider...." to "...the authenticated OAuth role." under the Examples heading. That info is now already covered at the top of the page.

    

- [x] Insert a heading that captures the purpose of this first example config (see below).

    

- [x] Then start the sub-section with "The following example shows how to create a basic configuration to run an OpenID Connect Provider. This is especially useful to use during development and testing prior to deploying to the cloud (where the application typically uses a cloud-based SSO provider instead). The example uses a simple basic registry, which is defined in the server configuration, to manage a single user:" and remove the paragraph "The following example assumes......with LDAP config.".

    

- [x] Make "basic registry" into a link to the Application Security feature (it contains examples and explanations of basic registry configurations).

    Then have the first example here.
    

- [x] Simplify the basicRegistry section of the example so that it only shows one user and no groups (as shown in the first example config in [https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#appSecurity-3.0.html](AppSecurity feature). We need this to be as minimal as possible; we don't need to show multiple users/groups here as that's not what we're demonstrating in this example.

    

- [x] Then after the example, explain what the example shows: "The example configuration uses localStore to store client data and token status, which means all client data and token status is held in memory. This is fine for test and development purposes but if you were to reconfigure the server, the localStore might be cleared so it is not suitable to use for production purposes when a database should be used instead. (then explain what the settings are doing and where those values come from - otherwise it just looks like a bunch of values plucked from the ether.)"

    

- [x] Then explain what the openidConnectProvider values show and where they come from.

    

- [x] I don't think we need the third example (lead-in says "Example to add clients with localstore:") because that exact example is already part of the larger first example, which is described as using localstore.

   

- [x]  Can we also @ayoho just move the following clientManager lines into the "data source with authentication" example instead of it standing alone as it's not really an alternative configuration?

    <clientManager>
       <user name="testuser" />
       <group name="oidcadmin" />
    </clientManager>

    Then briefly explain what it does.
    So then we just have two examples but they each need their own sub-heading that captures what they're demonstrating. eg "Managing users with basic registry" and "Managing users with an authenticated database" maybe? @ayoho
    

- [x] The indentations in the XML in the final example needs tidying up so that opening and closing tags match and everything's correctly nested. Also, the random > character should be on the previous line, not on a line on its own.

**Comment by lauracowen** [comment]
I think there's a build/merge glitch again - I don't see all of the changes above in the topic on the draft website. Can you take a look and track down your changes? https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#openidConnectServer-1.0.html

**Comment by lauracowen** [comment]
Sorry, I still don't see all your changes here https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/openidConnectServer-1.0.html

Can you go through each of the checked-off steps above and make sure they're visible in the built file? 

Also, can you check for broken links? With the new docs framework, they're shown in red text I think.

**Comment by ManasiGandhi** [comment]
@lauracowen here is the draft with the built changes https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/openidConnectServer-1.0.html. 

I went through each of the checked-off steps above and they're visible in the built file.

**Comment by lauracowen** [comment]
localstore example:
- "However, if you want to reconfigure the server" -> "However, if you reconfigure the server" (just a bit more direct about what happens)
- I think this explanation under the example needs re-ordering as it talks about databases in the localstore example section, whereas the next example section is what focuses on databases.
  - Move "In the example,..." paragraph to be the first paragraph after the example config. This is referring directly back to the example config.
  - Remove the paragraph "If you use a database,..." as that's not relevant to this section; that's about the next section.
  - Have "The example configuration uses localstore..." as the last paragraph in this example section.
  - Don't highlight "localstore" as this isn't referring to the element but to the thing itself - check the KC for whether it's spelt "localstore" or "local store" or whatever.
  - Talk to the SME and re-draft the paragraph "In the example,..." paragraph so that it's more explanatory. Just describing what's in the example config isn't that helpful - eg I can see from the XML example that the ID attribute is set to "true" but I don't understand why or why it's important enough for you to call it out here. The "signatureAlgorithm" sentence is fine. The keyStoreRef sentence doesn't explain what this means - what does this attribute do and why is it set to that? Ditto for the last sentence.

database example:
- Start the section with a variation on the bit you removed from the previous example: "When you use a database as a store, clients are not specified in the server.xml file. Instead, they are added to the database through the registration endpoint."
- Can you check what this second example shows? It doesn't look like JDBC config here (which is what the removed sentence from the previous section implied it should show) - it looks identical to the previous example's localstore configuration.

**Comment by ManasiGandhi** [comment]
@lauracowen I worked on your edit suggestions:


   

- [x]  "However, if you want to reconfigure the server" -> "However, if you reconfigure the server" (just a bit more direct about what happens)

    

- [x] I think this explanation under the example needs re-ordering as it talks about databases in the localstore example section, whereas the next example section is what focuses on databases.

        

- [x] Move "In the example,..." paragraph to be the first paragraph after the example config. This is referring directly back to the example config.

        

- [x] Remove the paragraph "If you use a database,..." as that's not relevant to this section; that's about the next section.

        

- [x] Have "The example configuration uses localstore..." as the last paragraph in this example section.

        

- [x] Don't highlight "localstore" as this isn't referring to the element but to the thing itself - check the KC for whether it's spelt "localstore" or "local store" or whatever.

        

- [x] Talk to the SME and re-draft the paragraph "In the example,..." paragraph so that it's more explanatory. Just describing what's in the example config isn't that helpful - eg I can see from the XML example that the ID attribute is set to "true" but I don't understand why or why it's important enough for you to call it out here. The "signatureAlgorithm" sentence is fine. The keyStoreRef sentence doesn't explain what this means - what does this attribute do and why is it set to that? Ditto for the last sentence. (updated per Adam Yoho's suggestions)

**Comment by lauracowen** [comment]
Thanks for making the paragraph under the first example more explanatory - that's much better.

I'm going to sign this off because it's fine in terms of content now. I spotted about three things that you should fix before passing it to Karen for editorial review though:

- [x] There's a syntax/highlighting error in the paragraph that starts "In the example, the openidConnectProvider element is configured..."
- [x] "The `jwkEnabled` attribute set to true indicates" - "true" should also be in backticks because it's a value in the config.
- [x] Should it be "localstore" or "local store"? Please check and make consistent throughout (it's not). Ask @chirp1  if you're not sure.

**Comment by ManasiGandhi** [comment]
@chirp1 Can you review this draft https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/reference/feature/openidConnectServer-1.0.html?

**Comment by ManasiGandhi** [comment]
@chirp1 Updated draft link https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/feature/openidConnectServer-1.0.html

**Comment by ManasiGandhi** [comment]
@chirp1 20.12 updated draft link https://draft-openlibertyio.mybluemix.net/docs/20.0.0.12/reference/feature/openidConnectServer-1.0.html.

**Comment by chirp1** [comment]
Hi Manasi,
Nice job on the configuration examples! I have a few comments:
- Move the "The example uses a basic registry ...." sentence immediately after the "The following example shows how to create ..." sentence as the thoughts in the two sentences are closely related. If necessary, make adjustments to the sentences to make them read better together.
.- We want to be consistent with terms in topics and across topics. You use the term "basic registry ", but it's not consistent with what we are using. Research what the term should be and update your term.
- For "that is set to true, indicates ", remove the comma.
- For "The example configuration uses local store to store...", change the wording so that you work in "localStore" so that you add concreteness to the sentence.
- For "...memory is not suitable to use for production purposes, ", remove "to use" to reduce wordiness.
- For "You can use a database instead of a local store in production. ", the use of "You can" in this case is really saying "You have the option to use a database." We don't want to imply an option because using a database in production is the correct thing to do. Reword by giving a command, In general, do not use "you can" when you don't want to give users a choice.
- Somewhere in your  "Managing OAuth clients with an authenticated database" section mention, mention that the database is for production purposes.
- Change "the user testuser or members " to "the testuser user or members ".

**Comment by ManasiGandhi** [comment]
I worked on Karen's review for this issue. Waiting on the changes to show after the builds are resolved.

**Comment by ManasiGandhi** [comment]
@chirp1 I worked on your review. Here is a link to the updated draft https://draft-openlibertyio.mybluemix.net/docs/21.0.0.7/reference/feature/openidConnectServer-1.0.html


    

- [x] Move the "The example uses a basic registry ...." sentence immediately after the "The following example shows how to create ..." sentence as the thoughts in the two sentences are closely related. If necessary, make adjustments to the sentences to make them read better together.

    

- [x] We want to be consistent with terms in topics and across topics. You use the term "basic registry ", but it's not consistent with what we are using. Research what the term should be and update your term.

    

- [x] For "that is set to true, indicates ", remove the comma.

    

- [x] For "The example configuration uses local store to store...", change the wording so that you work in "localStore" so that you add concreteness to the sentence.

    

- [x] For "...memory is not suitable to use for production purposes, ", remove "to use" to reduce wordiness.

    

- [x] For "You can use a database instead of a local store in production. ", the use of "You can" in this case is really saying "You have the option to use a database." We don't want to imply an option because using a database in production is the correct thing to do. Reword by giving a command, In general, do not use "you can" when you don't want to give users a choice.

    

- [x] Somewhere in your "Managing OAuth clients with an authenticated database" section mention, mention that the database is for production purposes.

    

- [x] Change "the user testuser or members " to "the testuser user or members ".

**Comment by dmuelle** [comment]
First example looks good but still need to update the `Add users and groups to the clientManager role in production` example. Note that in the previous example, it says:

> However, storing client data and token status in memory is not suitable for production purposes, as reconfiguring the server might clear the local store. Use a database instead of a local store in production.

Then in the intro to the "production" example, it says

> The following example uses a uses localStore to store the clients for production purposes:

Which we just advised not do. Need to resolve this contradiction and add more detail to the production example. Recommend having an SME look over it when the edits are ready

**Comment by ManasiGandhi** [comment]
@ayoho I made the updates per your suggestion https://draft-openlibertyio.mybluemix.net/docs/21.0.0.10/reference/feature/openidConnectServer-1.0.html Could you review the draft for the updates?

**Comment by ManasiGandhi** [comment]
Adam's comment from slack,

"Changes look good"

**Comment by dmuelle** [comment]
Hi Manasi- the examples look good. Just a couple minor recommendations:

With this configuration, users need to authenticate once to access the Open Liberty resources such as HTML, JavaServer Pages (JSP) files, and servlets.
--->
With this configuration, users need to authenticate only once to access Open Liberty resources such as HTML, JavaServer Pages (JSP) files, and servlets.

Use a database instead of a local store in production.
--->
In production environments, client data and token status are stored in a database instead of a local store.

The OpenID Connect provider generates a JSON Web Key by using the certificate in the key store...
--->
The OpenID Connect provider generates a JSON Web Key by using the certificate in the keystore...

**Comment by ManasiGandhi** [comment]
https://docs-staging-openlibertyio.mybluemix.net/docs/21.0.0.10/reference/feature/openidConnectServer-1.0.html
